pkg/cmd/ssh/command.go

```diff
@@ -0,0 +1,460 @@
+	cmd.Stdout = o.IOStreams.Out
+	cmd.Stdin = o.IOStreams.In
+	cmd.Stderr = o.IOStreams.ErrOut
+
+	return cmd.Run()
+}
+
+// waitForSignal informs the user about their options and keeps the
+// bastion alive until gardenctl exits.
+func waitForSignal(o *Options, bastion *operationsv1alpha1.Bastion, nodePrivateKeyFile string, node *corev1.Node) error {
+	nodeHostname := "SHOOT_NODE"
+
+	if node != nil {
+		var err error
+
+		nodeHostname, err = getNodeHostname(node)
+		if err != nil {
+			return err
+		}
+	}
+
+	bastionAddr := preferedBastionAddress(bastion)
+	connectCmd := sshCommandLine(o, bastionAddr, nodePrivateKeyFile, nodeHostname)
+
+	fmt.Fprintln(o.IOStreams.Out, "Connect to Shoot nodes by using the bastion as a proxy/jump host, for example:")
+	fmt.Fprintln(o.IOStreams.Out, "")
+	fmt.Fprintln(o.IOStreams.Out, connectCmd)
+	fmt.Fprintln(o.IOStreams.Out, "")
+	fmt.Fprintln(o.IOStreams.Out, "Press Ctrl-C to stop gardenctl, after which the bastion will be removed.")
+
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, os.Interrupt)
+	<-signalChan
+
+	return nil
+}
+
+func sshCommandLine(o *Options, bastionAddr string, nodePrivateKeyFile string, nodeName string) string {
+	proxyPrivateKeyFlag := ""
+	if o.SSHPrivateKeyFile != "" {
+		proxyPrivateKeyFlag = fmt.Sprintf(" -o IdentitiesOnly=yes -i %s", o.SSHPrivateKeyFile)
+	}
+
+	proxyCmd := fmt.Sprintf(
+		"ssh -W%%h:%%p -o StrictHostKeyChecking=no%s %s@%s",
+		proxyPrivateKeyFlag,
+		SSHBastionUsername,
+		bastionAddr,
+	)
+
+	connectCmd := fmt.Sprintf(
+		`ssh -o "StrictHostKeyChecking=no" -o "IdentitiesOnly=yes" -i %s -o "ProxyCommand=%s" %s@%s`,
+		nodePrivateKeyFile,
+		proxyCmd,
+		SSHNodeUsername,
+		nodeName,
+	)
+
+	return connectCmd
+}
+
+func keepBastionAlive(ctx context.Context, gardenClient client.Client, bastion *operationsv1alpha1.Bastion, stderr io.Writer) {
+	ticker := time.NewTicker(3 * time.Minute)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-ticker.C:
+			// re-fetch current bastion
+			key := types.NamespacedName{Name: bastion.Name, Namespace: bastion.Namespace}
+
+			// reset annotations so that we fetch the actual current state
+			bastion.Annotations = map[string]string{}
+
+			if err := gardenClient.Get(ctx, key, bastion); err != nil {
+				fmt.Fprintf(stderr, "Failed to keep bastion alive: %v\n", err)
+			}
+
+			// add the keepalive annotation
+			oldBastion := bastion.DeepCopy()
+			bastion.Annotations[v1beta1constants.GardenerOperation] = v1beta1constants.GardenerOperationKeepalive
+
+			if err := gardenClient.Patch(ctx, bastion, client.MergeFrom(oldBastion)); err != nil {
+				fmt.Fprintf(stderr, "Failed to keep bastion alive: %v\n", err)
+			}
+		}
+	}
+}
+
+func getShootNodePrivateKey(ctx context.Context, gardenClient client.Client, shoot *gardencorev1beta1.Shoot) ([]byte, error) {
+	secret := &corev1.Secret{}
+	key := types.NamespacedName{
+		Name:      fmt.Sprintf("%s.ssh-keypair", shoot.Name),
```

Comment:
use `ShootProjectSecretSuffixSSHKeypair` constant?

Reply:
We cannot use that package because once we import it, we get

> ../../../../pkg/mod/github.com/gardener/gardener@v1.23.0/pkg/client/kubernetes/runtime_client.go:84:9: undefined: manager.NewClientBuilder

As mentioned in another comment, we probably need to wait for ctrlruntime 0.9 support in Gardener.

I can leave a TODO comment though.

Reply:
ok then add a TODO comment